### ISSUE TITLE
Security concern

### ISSUE THREAD

**Issue by zidingz** [nodejs/node-gyp#2559]
Hey there!

I belong to an open source security research community, and a member (@dwisiswant0) has found an issue, but doesn’t know the best way to disclose it.

If not a hassle, might you kindly add a `SECURITY.md` file with an email, or another contact method? GitHub [recommends](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository) this best practice to ensure security issues are responsibly disclosed, and it would serve as a simple instruction for security researchers in the future.

Thank you for your consideration, and I look forward to hearing from you!

(cc @huntr-helper)

**Comment by rvagg** [comment]
yeah, fair, re SECURITY.md, we should probably add one, but I think it would mostly just point people to follow the process in https://github.com/nodejs/node/blob/master/SECURITY.md - so for now you could follow this.

**Comment by dwisiswant0** [comment]
> yeah, fair, re SECURITY.md, we should probably add one, but I think it would mostly just point people to follow the process in https://github.com/nodejs/node/blob/master/SECURITY.md - so for now you could follow this.

Regarding that, I have reported through HackerOne. But unfortunately marked as not applicable because it is out-of-scope.

**Comment by rvagg** [comment]
@Trott @mcollina I don't think I have access to hackerone anymore, so I'm going to trust that any rejection is appropriate considering we have responsibility for node-gyp in the org.

**Comment by Trott** [comment]
> @Trott @mcollina I don't think I have access to hackerone anymore, so I'm going to trust that any rejection is appropriate considering we have responsibility for node-gyp in the org.

Looping in @vdeturckheim who triaged that particular item.

Vladimir can confirm, but I believe the conclusion was that while the bug may be real, there's no way for it to impact Node.js end-users. So it can be reported here instead of in HackerOne. I'd want Vladimir to confirm though before actually reporting details.

**Comment by Trott** [comment]
@rvagg I've emailed you more details. I think the issue is OK to discuss openly in the issue tracker but I'll wait for your and Vladimir's agreement on that.

**Comment by rvagg** [comment]
yep, as per email I'm +1 on having this discussed in open on the nodejs/gyp-next repo

**Comment by vdeturckheim** [comment]
> Vladimir can confirm, but I believe the conclusion was that while the bug may be real, there's no way for it to impact Node.js end-users. So it can be reported here instead of in HackerOne. I'd want Vladimir to confirm though before actually reporting details.

Yes, I could not see any way this could impact any Node.js end user.

> @rvagg I've emailed you more details. I think the issue is OK to discuss openly in the issue tracker but I'll wait for your and Vladimir's agreement on that.

Answering on the email thread too, but yes let's go in the open :)

**Comment by dwisiswant0** [comment]
> @rvagg I've emailed you more details. I think the issue is OK to discuss openly in the issue tracker but I'll wait for your and Vladimir's agreement on that.

> Answering on the email thread too, but yes let's go in the open :)

@rvagg @Trott @vdeturckheim I'm afraid the details have already been explained through the [huntr](https://huntr.dev) platform privately. The associated repository maintainers should be able to review via this URL: https://huntr.dev/bounties/3ce8dfe5-6c38-4819-a8e2-4baa6e8f9c62/.

Correct me if I'm wrong, @zidingz @huntr-helper. :)

**Comment by vdeturckheim** [comment]
I don't believe I have access to huntr

**Comment by mcollina** [comment]
I do not support huntr in any form, the official security tracker is on HackerOne as detailed in the org security channel.

**Comment by Trott** [comment]
@dwisiswant0 Is it safe for me to interpret your comment as "The issue has been explained to the appropriate Node.js parties in their HackerOne tracker and they are free to disclose it publicly if they choose."?

If so, I'll happily share the details here.

**Comment by Trott** [comment]
I've logged into huntr and can see that the report there is effectively identical to what we received in HackerOne.

**Comment by Trott** [comment]
(I guess I can mark the huntr report as invalid and then disclose?)

**Comment by dwisiswant0** [comment]
I'm up with you, @Trott.

**Comment by Trott** [comment]
Marked as Invalid in Huntr. Here's the gist of the report:

This line: https://github.com/nodejs/node-gyp/blob/0da2e0140dbf4296d5ba7ea498fac05e74bb4bbb/gyp/tools/pretty_gyp.py#L93

...contains a regex that is vulnerable to backtracking issues.

```python
import re
closing_prefix_re = re.compile(r"(.*?[^\s\]\}\)]+.*?)([\]\}\)],?)\s*$")
closing_prefix_re.search('0' * 3456)
```

Options to address this issue are:

* Ignore it. If you want to DoS yourself in your build tool, cool. It's not a security issue.
* Remove `pretty_gyp.py` entirely. Not sure anyone actually uses it.
* Replace the regex with a regex that does not have this issue or with a non-regex approach.

**Comment by Trott** [comment]
@rvagg Are there tests for `pretty_gyp.py`? That regex seems needlessly complex. It doesn't actually use the capture groups in the code, for example. I think it can be, uh, "simplified" to `/[^\s\]\}\)]\s*[\]\}\)]+,?\s*$/`.

I'm not 100% sure that isn't still susceptible to backtracking issues, but if it is, it is less obviously so, at least to me.

@meekdenzo Because I know you already have @davisjam's tools set up for this stuff, can I ask you for a favor and have you run `/[^\s\]\}\)]\s*[\]\}\)]+,?\s*$/` through the Perl script and see if it gets reported as a problematic regex or not?

**Comment by meekdenzo** [comment]
@Trott `/[^\s\]\}\)]\s*[\]\}\)]+,?\s*$/ ` is reported not vulnerable.

**Comment by davisjam** [comment]
Looks ok to me at a quick glance
________________________________

**Comment by richardlau** [comment]
`pretty_gyp.gyp` comes from https://github.com/nodejs/gyp-next -- this issue should probably be moved there.
